modules/rtbhouseBidAdapter.js

```diff
@@ -96,30 +63,257 @@ export const spec = {
+  return {
+    id: slot.bidId,
+    banner: mapBanner(slot),
+    native: mapNative(slot),
+    tagid: slot.adUnitCode.toString()
+  };
+}
+
+/**
+ * Produces OpenRTB Banner object
+ */
+function mapBanner(slot) {
+  if (slot.mediaType === 'banner' ||
+    utils.deepAccess(slot, 'mediaTypes.banner') ||
+    (!slot.mediaType && !slot.mediaTypes)) {
+    return {
+      w: slot.sizes[0][0],
+      h: slot.sizes[0][1],
+      format: slot.sizes.map(size => ({
+        w: size[0],
+        h: size[1]
+      }))
+    };
+  }
+}
+
+/**
+ * Produces an OpenRTB Site object
+ */
+function mapSite(validRequest) {
+  const pubId = validRequest && validRequest.length > 0
+    ? validRequest[0].params.publisherId : 'unknown';
+  return {
+    publisher: {
+      id: pubId.toString(),
+    },
+    page: utils.getTopWindowUrl(),
+    name: utils.getOrigin()
+  }
+}
+
+/**
+ * Produces an OpenRTB Native object from a slot config
+ */
+function mapNative(slot) {
+  if (slot.mediaType === 'native' || utils.deepAccess(slot, 'mediaTypes.native')) {
+    return {
+      request: {
+        assets: mapNativeAssets(slot)
+      },
+      ver: '1.1'
+    }
+  }
+}
+
+/**
+ * Produces an OpenRTB Native Ads Assets objects
+ */
+function mapNativeAssets(slot) {
+  const params = slot.nativeParams || utils.deepAccess(slot, 'mediaTypes.native');
+  const assets = [];
+  if (params.title) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.TITLE,
+      required: params.title.required ? 1 : 0,
+      title: {
+        len: params.title.len || 140
+      }
+    })
+  }
+  if (params.image) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.IMAGE,
+      required: params.image.required ? 1 : 0,
+      img: mapNativeImage(params.image, OPENRTB.NATIVE.IMAGE_TYPE.MAIN)
+    })
+  }
+  if (params.icon) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.ICON,
+      required: params.icon.required ? 1 : 0,
+      img: mapNativeImage(params.icon, OPENRTB.NATIVE.IMAGE_TYPE.ICON)
+    })
+  }
+  if (params.sponsoredBy) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.SPONSORED,
+      required: params.sponsoredBy.required ? 1 : 0,
+      data: {
+        type: OPENRTB.NATIVE.DATA_ASSET_TYPE.SPONSORED
+      }
+    })
+  }
+  if (params.body) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.BODY,
+      required: params.body.request ? 1 : 0,
+      data: {
+        type: OPENRTB.NATIVE.DATA_ASSET_TYPE.DESC,
+        len: params.body.len || undefined
+      }
+    })
+  }
+  if (params.clickUrl) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.CLICK_URL,
+      required: params.clickUrl.required ? 1 : 0,
+      data: {
+        type: OPENRTB.NATIVE.DATA_ASSET_TYPE.CLICK_URL
+      }
+    })
+  }
+  if (params.cta) {
+    assets.push({
+      id: OPENRTB.NATIVE.ASSET_ID.CTA,
+      required: params.cta.required ? 1 : 0,
+      data: {
+        type: OPENRTB.NATIVE.DATA_ASSET_TYPE.CTA_TEXT,
+        len: params.cta.len || undefined
```

Comment:
We don't want undefined in any case. This should be treated as an error that we want to get raised :duck:

Reply:
But it's optional field accord to OpenRTB and if it's `undefined` it won't be send in request, right? Anyway I will omit this ` || undefined` because it looks bad, but raising exception has no sense also.

Reply:
But if params.cta is there should the other props specific to CTA be also provided, or correctly defaulted?